language_service/src/state/tests.rs

```diff
@@ -924,6 +936,207 @@ fn close_notebook_clears_errors() {
     );
 }
 
+#[tokio::test]
+async fn update_notebook_with_valid_dependencies() {
+    let fs = FsNode::Dir(
+        [dir(
+            "project",
+            [
+                file("qsharp.json", r#"{ }"#),
+                dir(
+                    "src",
+                    [file(
+                        "file.qs",
+                        r#"namespace Foo { function Bar() : Unit { } }"#,
+                    )],
+                ),
+            ],
+        )]
+        .into_iter()
+        .collect(),
+    );
+
+    let fs = Rc::new(RefCell::new(fs));
+    let errors = RefCell::new(Vec::new());
+    let mut updater = new_updater_with_file_system(&errors, &fs);
+
+    updater
+        .update_notebook_document(
+            "notebook.ipynb",
+            &NotebookMetadata {
+                target_profile: None,
+                language_features: LanguageFeatures::default(),
+                manifest: None,
+                project_root: Some("project".to_string()),
+            },
+            [("cell1", 1, "open Foo;Bar();")].into_iter(),
+        )
+        .await;
+
+    expect_errors(
+        &errors,
+        &expect![[r#"
+        []
+    "#]],
+    );
+}
+
+#[tokio::test]
+async fn update_notebook_reports_errors_from_dependencies() {
```

Comment:
Is this test saying that if there are errors in a dependency, we'd see those errors when we open the associated namespace in the notebook?

Reply:
Maybe I'm not understanding the relationship between the project and the notebook in these tests. It doesn't seem right to me that people would get errors from external code like that.

Reply:
If a dependency can't compile, the user code can't compile either. Since dependencies are brought in as source and compiled here rather than distributed as precompiled binaries (the same mechanism as Rust's source distribution), all dependencies must be compiled in the right order before the user code. So yes, it's expected and necessary for the user to see failures from dependencies to know there is a bug or misconfiguration in one of them that will prevent any other code from working. These errors show up in the problems view in VS Code, and clicking them will jump to the offended dependency file, which can help with reporting bugs in published code.

One thing I should do here is mimic the pattern Alex set up that does a best effort compile of the user code even when dependencies fail... it will show errors on everything the user code tries to pull from that or any dependency, but that can be a helpful signal to the user that things have gone wrong.

Reply:
Okay, I can see the reasoning behind that. Thanks for the explanation!